man/qvwm.man
.de EX
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..

.TH QVWM 1x "1 April 2006" "Version 1.1.12-8-IKu"

.SH NAME
qvwm \- Windows 95/98 like window manager for X11

.SH SYNPSIS
.B qvwm
[ \fIoptions\fP ]

.SH DESCRIPTION

Qvwm is a simple, lightweight Windows 95-like window manager for X11.  It
looks much like Windows 95/98, and provides the functions of start menu,
taskbar, task switcher, configurable keyboard shortcuts and an imitation of
desktop shortcuts.  Also, it supports virtual screen and a pager functions,
which are common for X window managers. Furthermore, it deals with any
languages because it is considered with internationalization.

Qvwm is written in C++, without depending on twm or fvwm.  

.SH OPTIONS
Qvwm accepts the following options on the command line:

.TP
\fB--clientid\fP \fIclient_id\fP
Specifies the client Id from the previous session.
It is equivalent to \fB-id\fP or \fB-clientId\fP.
.TP
\fB--config\fP
Displays compile options.
It is equivalent to \fB-cfg\fP.
.TP
\fB--display\fP \fIdisplay_name\fP
Specifies the display name that qvwm uses by display_name. If you don't
specify this option, qvwm uses the environment variable $DISPLAY.
It is equivalent to \fB-d\fP or \fB-display\fP.
.TP
\fB--help\fP
Displays simple description of qvwm command line options.
.TP
\fB--noparse\fP
Starts qvwm without reading any configuration files.
It is equivalent to \fB-n\fP.
.TP
\fB--rcfile\fP \fIrcfile_name\fP
Specifies the configuration file that qvwm reads on startup by rcfile_name. 
By default, $HOME/.qvwmrc is read; if it doesn't exist, 
<QVWMDIR>/system.qvwmrc.
It is equivalent to \fB-f\fP.
.TP
\fB--restart\fP
Starts qvwm without executing a [Startup] section in a configuration file.
It is equivalent to \fB-r\fP.
.TP
\fB--synch\fP
Executes qvwm on synchronization mode.  Although qvwm with this option 
works slowly, qvwm may work stably.
It is equivalent to \fB-s\fP or \fB-synch\fP.
.TP
\fB--version\fP
Displays the version and copyright of qvwm.
It is equivalent to \fB-v\fP.


.SH VARIABLE CONFIGURATION
The following are variables you specify in the configuration file's
[Variables] section. The values in parentheses are default.
.TP
\fBAllowRemoteCmd\fP (False)
Sets whether qvwm allows other processes of the user that starts qvwm
to send internal commands to it.
.TP
\fBAutoRaise\fP (False)
Sets whether the active window raises on top automatically when ClickToFocus
is False.
.TP
\fBAutoRaiseDelay\fP (500)
Sets the delay time taken to raise the active window automatically by the
millisecond. (more than 0)
.TP
\fBButtonColor\fP (gray)
Sets the button color.
.TP
\fBButtonStringColor\fP (black)
Sets the button string color.
.TP
\fBCascadeMenuFont\fP (null)
Sets the font for the cascade memu, that is, child menus of the start menu.
If this variable is not specified, DefaultFont is used.
.TP
\fBClickToFocus\fP (True)
If True, qvwm activates the window you click. Otherwise, the active window
changes according to the movement of mouse cursor.
.TP
\fBClickingRaises\fP (True)
If True, a window is raised even if you click the client area.
When ClickToFocus is True, ClickingRaises is always True.
.TP
\fBClockFormat\fP ("%R")
Sets the format of the taskbar clock.  The format follows strftime(3).
.TP
\fBClockLocaleName\fP ("")
Sets the locale of taskbar clock.
(see /usr/X11R6/lib/X11/locale/locale.alias)
.TP
\fBClockMessageFormat\fP ("%A, %B %e, %Y")
Sets the format of the message of the taskbar clock tooltip.  The
format follows strftime(3).
.TP
\fBClockStringColor\fP (black)
Sets the taskbar clock string color.
.TP
\fBCloseButtonMessage\fP ("Close")
Sets the message of the close button tooltip.
.TP
\fBCloseSound\fP (NULL)
Sets the sound file played back when a window is closed.
.TP
\fBCtrlMenuFont\fP (null)
Sets the font for the control memu.  If this variable is not specified,
DefaultFont is used.
.TP
\fBCtrlMoveRatio\fP (1.0)
Sets the ratio to the normal movement (13 dots) when you move a window
using the keyboard pressing a control key.
.TP
\fBCursorColor\fP (white)
Sets the cursor color.
.TP
\fBDefaultFolderIcon\fP (null)
Sets the default icon for menu items, which open submenus.  If you do not
specify, it will be a little 16x16 folder picture.
.TP
\fBDefaultFont\fP ("-*-*-medium-r-normal-*-14-*")
Sets the default font.
.TP
\fBDefaultIcon\fP (null)
Sets the default image of an icon used in titlebars and taskbar buttons.  The
size must be 16x16.  If this variable is not specified, a default logo is
used.  (Currently, it is the Debian swirl logo, but can be changed by
editing the source src/bitmaps/logo16.xpm file.)
.TP
\fBDefaultLargeIcon\fP (null)
Sets the default image of an icon used in the task switcher.  The
size must be 32x32.  If this variable is not specified, the built-in
Windows logo is used.
.TP
\fBDefaultMenuItemIcon\fP (null)
Sets the default icon for normal menu items, ie for this items, which run
programs or execute commands.  No icon is used, if you do not specify it here,
nor in your menu specifications.
.TP
\fBDefaultShortcutIcon\fP (null)
Sets the default image of a shortcut icon.  The size must be 32x32.
If this variable is not specified, the built-in image is used.
.TP
\fBDesktopActiveColor\fP (blue)
Sets the desktop active color for the selected shortcut icon.
.TP
\fBDesktopColor\fP (royalblue)
Sets the desktop color.  If you set WallPaper to Windows2000, the
desktop is filled with this color.
.TP
\fBDialogColor\fP (gray)
Sets the exit dialog color.
.TP
\fBDialogFont\fP (null)
Sets the font for the exit dialog.  If this variable is not specified,
DefaultFont is used.
.TP
\fBDialogImage\fP ("")
Sets the exit dialog background image.
.TP
\fBDialogStringColor\fP (black)
Sets the exit dialog string color.
.TP
\fBDisableDesktopChange\fP (False)
Sets whether the user cannot move or resize the taskbar, the pager,
and icons.
.TP
\fBDisableTaskbarDragging\fP (True)
Forbids dragging the taskbar with mouse.  Set to False to enable dragging.
.TP
\fBDoubleClickRange\fP (5)
Sets the max distance between two clicks for double click by the dot. 
(more than 1)
.TP
\fBDoubleClickTime\fP (400)
Sets the max interval between two clicks for double click by the millisecond.
(more than 1)
.TP
\fBEdgeResistance\fP (0)
Sets the resistance needed for moving the frame of a window over the
edges of the screen.
.TP
\fBEnableAlsa\fP (True)
Sets whether you enable ALSA (Advanced Linux Sound Architecture) if OS
supports it.
.TP
\fBEnableEsd\fP (True)
Sets whether you enable EsounD if installed.
.TP
\fBEnableSound\fP (True)
Sets whether you enable the sound facility.
.TP
\fBExpandSound\fP (NULL)
Sets the sound file played back when a window is expanded by
QVWM_EXPAND_?? functions.
.TP
\fBFocusOnMap\fP (False)
Sets whether a newly mapped window is activated automatically when
ClickToFocus is False.
.TP
\fBFrameActiveColor\fP (gray)
Sets the window frame active color.
.TP
\fBFrameActiveImage\fP ("")
Sets the window frame active background image.
.TP
\fBFrameColor\fP (gray)
Sets the window frame inactive color.
.TP
\fBFrameImage\fP ("")
Sets the window frame inactive background image.
.TP
\fBFrameBorderWidth\fP (2)
Sets the border width of a window frame, excluding shade width.
.TP
\fBFrameTitleHeight\fP (18)
Sets the titlebar height of a window frame.
.TP
\fBFullOpaque\fP (True)
Sets whether the contents of a window are redrawn on resizing.
.TP

CHANGES

Qvwm source version 1.1.12 
http://ftp.debian.org/debian/pool/main/q/qvwm/qvwm_1.1.12.orig.tar.gz

plus patches to version 1.1.12-3 from Debian.org
http://ftp.debian.org/debian/pool/main/q/qvwm/qvwm_1.1.12-3.diff.gz

(downloaded on 2006-03-16)


2006-03-16  Ivan Kurmanov, http://ahinea.com/en/

* Fixed: window and taskbar button naming is now based on
  WM_NAME window property primarily and WM_ICON_NAME as the
  second best.  This avoids "Untitled" name for some
  applications/windows.

* Added restart item to the default start menu
 
* Two new configuration variables for menu look preferences:
  DefaultMenuItemIcon and DefaultFolderIcon

* New configuration variable: DisableTaskbarDragging.  Should speak for
  itself.


2006-03-17  Ivan Kurmanov

* Window maximization is now more intuitive: if a window is maximized already,
  it is restored when you press Alt-Enter.

  (QVWM_MAXIMIZE function is now toggling the maximized state, so for an
  already maximized window it does Restore instead.  There is also new
  QVWM_MAXIMIZE_ONEWAY function for the old, unconditional behaviour.)

* Respect INIT_MAXIMIZE and INIT_MINIMIZE window options (as
  configured in .qvwmrc) on qvwm restart

* Startmenu's logo (which is usually a vertically-laid text) is now optional

* Internal: CreateImageFromFile() now does not choke on an empty string


2006-03-19  Ivan Kurmanov

* Removed useless shit lines from the taskbar (between the start button and
  the task buttons, and between the task buttons and the system tray area)

* Menu item icons are now centered in their little cells.
  For sub-menus a little folder icon is now hard-coded
  default, but there is no longer a default for non-folder
  menu items.  (Previously there was an ugly X icon default.)

* MenuItemMinimalHeight configuration variable.  Sets the default height of
  the menu items.  Default value: 18;

* Left margin fixed for menu items (it was too big)


2006-03-20  Ivan Kurmanov

* Fixed tooltip flicker, which happened sometimes at screen bottom.

* Menu drawing-padding (left margin) and positioning
  improved.  Care taken to avoid putting a menu out of the
  visible screen boundary, if possible.  Menu items' height
  now depends on the font used, but MenuItemMinimalHeight
  variable is still respected.

* Two new externally-accessible functions for window
  management: QVWM_FOCUS and QVWM_RAISE_FOCUS; via remote
  command interface they accept a window name parameter.


2006-03-21

* Reimplemented task switcher layout.  It is now vertical,
  like a menu.  

  In case, when there are lots of windows open, it will
  scroll vertically as you go through the window list.  It
  will keep the appropriate part of it visible.

  Also, system does not longer freeze while you switch
  tasks.

* Fixed a bug which stopped qvwm from reading the .qvwmrc
  file on restart, if it had found a syntax error there in
  the past.  So, even if you had corrected the rc file, you
  still were forced to quit and start it.  Not anymore.

* Added xterm to the default hard-coded start menu.  (You'll
  be able to start xterm if qvwm failed to parse its config
  file.)

* Disabling and enabling taskbar now changes maximum window
  sizes accordingly (but not immediately).


2006-03-25 released qvwm 1.1.12-6-IKu

* Fixed a little menu drawing glitch: checked mark (a tick)
  and selected mark (a bullet) were not aligned with the
  item name text.  (Would only be noticible with a
  significantly larger than default MenuItemMinimalHeight
  value.)

2006-03-28 released qvwm 1.1.12-7-IKu 

* Now will try harder to find a suitable or any at all font

src/qvwm.cc
/*
 * qvwm.cc
 *
 * Copyright (C) 1995-2001 Kenichi Kourai
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with qvwm; see the file COPYING.  If not, write to
 * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdio.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xresource.h>
#ifdef USE_SHAPE
#include <X11/extensions/shape.h>
#endif
#include <X11/Xatom.h>
#include "main.h"
#include "misc.h"
#include "qvwm.h"
#include "util.h"
#include "fbutton.h"
#include "tbutton.h"
#include "sbutton.h"
#include "menu.h"
#include "startmenu.h"
#include "event.h"
#include "taskbar.h"
#include "icon.h"
#include "qvwmrc.h"
#include "switcher.h"
#include "paging.h"
#include "pager.h"
#include "mini.h"
#include "key.h"
#include "indicator.h"
#include "focus_mgr.h"
#include "desktop.h"
#include "mwm.h"
#include "gnome.h"
#include "exec.h"
#include "image.h"
#include "session.h"
#include "tooltip.h"
#include "screen_saver.h"

Bool		Qvwm::umReserved = False;
XContext	Qvwm::context;

int Qvwm::TITLE_HEIGHT;
int Qvwm::BORDER_WIDTH;
int Qvwm::TOP_BORDER;

int Qvwm::BUTTON_WIDTH = 16;      // frame button width
int Qvwm::BUTTON_HEIGHT = 14;     // frame button height

/*
 * Qvwm class constructor --
 *   Position the window at that point if usrPos is True;
 *   otherwise, the position of the window obeys a given rule.
 */
Qvwm::Qvwm(Window w, Bool usrPos)
: wOrig(w)
{
  flags = 0;
  status = 0;
  mapped = False;
  imgLarge = NULL;
  imgSmall = NULL;
  nCmapWins = 0;
  name = shortname = NULL;
  isFirstMapped = True;
#ifdef USE_XSMP
  smClientId = NULL;
  wmRole = NULL;
#endif

  XGrabServer(display);

  /*
   * Get some hints.
   */
  wmHints = XGetWMHints(display, w);
  GetWindowSizeHints();
  Gnome::GetHints(this);
  Mwm::GetHints(this);
  GetWindowClassHints();
  FetchWMProtocols();
  GetTransientForHint();

  Gnome::SetInitStates(this);

  GetWindowAttrs(usrPos);

  if (Intersect(rc, rootQvwm->GetRect()))
    Gnome::ResetState(this, WIN_STATE_HID_WORKSPACE);
  else
    Gnome::SetState(this, WIN_STATE_HID_WORKSPACE);

  Rect rcCtrl, rcButton[3];

  CalcFramePos(rc, &rcParent, &rcTitle, rcCorner, rcSide, rcEdge, &rcCtrl,
	       rcButton);

  CreateFrame(rc);
  CreateParent(rcParent);
  CreateSides(rcSide);
  CreateCorners(rcCorner);  
  CreateEdges(rcEdge);
  CreateTitle(rcTitle);
  CreateCtrlButton(rcCtrl);

  if (CheckFlags(THIN_BORDER))
    XSetWindowBorderWidth(display, frame, 1);

  /*
   * Create three kinds of frame button.
   */
  fButton[0] = new FrameButton1(this, title, rcButton[0]);
  fButton[0]->ChangeImage(FrameButton::MINIMIZE);
  fButton[1] = new FrameButton2(this, title, rcButton[1]);
  fButton[1]->ChangeImage(FrameButton::MAXIMIZE);
  fButton[2] = new FrameButton3(this, title, rcButton[2]);
  fButton[2]->ChangeImage(FrameButton::CLOSE);
  
  if (imgLarge == NULL)
    imgLarge = GetFixedIcon(32);
  if (imgSmall == NULL)
    imgSmall = GetFixedIcon(16);

  Rect rect(-50, -50, 50, 50);
  QvImage* img = imgSmall->Duplicate();
  tButton = new TaskbarButton(this, rect, img);

  ctrlMenu = ::ctrlMenu;

  ChangeOrigWindow();
  SendConfigureEvent();

#ifdef USE_SHAPE
  int bShaped, cShaped;
  int xbs, ybs, xcs, ycs;
  unsigned int wbs, hbs, wcs, hcs;

  if (shapeSupport) {
    XShapeSelectInput(display, w, ShapeNotifyMask);
    XShapeQueryExtents(display, w, &bShaped, &xbs, &ybs, &wbs, &hbs,
		       &cShaped, &xcs, &ycs, &wcs, &hcs);
    if (bShaped) {
      SetFlags(SHAPED);
      SetShape();
    }
  }
#endif // USE_SHAPE

  if (UsePager)
    mini = new Miniature(this, rc);

  scKey->GrabKeys(frame);

  SetTransient();

  XUngrabServer(display);
}

/*
 * Qvwm class constructor --
 *   For only root window.
 */
Qvwm::Qvwm(Window root)
: wOrig(root), frame(root)
{
  Rect rect(0, 0, BUTTON_WIDTH, BUTTON_HEIGHT);

  rc = Rect(0, 0,
	    DisplayWidth(display, screen), DisplayHeight(display, screen));

  parent = title = ctrl = None;
  for (int i = 0; i < 4; i++) {
    corner[i] = None;
    side[i]= None;
  }
  mapped = True;
  flags = 0L;
  status = 0;
  nCmapWins = 0;
  name = new char[10];
  strcpy(name, "qvwm root");
  shortname = NULL;
  toolTip = NULL;

#ifdef USE_XSMP
  smClientId = NULL;
  wmRole = NULL;
#endif

  XRectangle ink, log;

  XmbTextExtents(fsTaskbar, StartButtonTitle, strlen(StartButtonTitle),
                 &ink, &log);
  log.width += 2 + StartButton::SYMBOL_SIZE + 4 + 2 + 4;

  taskBar = new Taskbar(this, log.width + 4, TaskbarRows);
  tButton = new StartButton(this, log.width);

  rect = Rect(0, 0, 0, 0);
  ctrlMenu = new Menu(DesktopMenuItem, fsCtrlMenu, NULL, this);

  XSaveContext(display, root, context, (caddr_t)this);

  focusQvwm = this;

  scKey->GrabKeys(root);
}

Qvwm::~Qvwm()
{
  int i;

  if (this != rootQvwm) {
    XDeleteContext(display, frame, context);
    XDeleteContext(display, parent, context);
    XDeleteContext(display, wOrig, context);
    XDeleteContext(display, title, context);
    for (i = 0; i < 4; i++) {
      XDeleteContext(display, corner[i], context);
      XDeleteContext(display, side[i], context);
    }
    XDeleteContext(display, ctrl, context);

    QvImage::Destroy(imgLarge);
    QvImage::Destroy(imgSmall);

    if (TitlebarImage)
      QvImage::Destroy(imgTitle);
    if (TitlebarActiveImage)
      QvImage::Destroy(imgActiveTitle);

    if (FrameImage) {
      for (i = 0; i < 4; i++) {
	QvImage::Destroy(imgSide[i]);
	QvImage::Destroy(imgCorner[i]);
      }
    }
    if (FrameActiveImage) {
      for (i = 0; i < 4; i++) {
	QvImage::Destroy(imgActiveSide[i]);
	QvImage::Destroy(imgActiveCorner[i]);
      }
    }

    if (GradTitlebar) {
      if (pixGrad)
	XFreePixmap(display, pixGrad);
      if (pixActiveGrad)
	XFreePixmap(display, pixActiveGrad);
    }

    delete toolTip;

    if (wmHints)
      XFree(wmHints);

    if (classHints.res_name)
      XFree(classHints.res_name);
    if (classHints.res_class)
      XFree(classHints.res_class);

#ifdef USE_XSMP
    if (wmRole)
      XFree(wmRole);
    if (smClientId)
      XFree(smClientId);
#endif

    /*
     * Delete frame buttons before the parent window(frame) is destroyed.
     */
    for (i = 0; i < 3; i++)
      delete fButton[i];
    delete tButton;

    /*
     * When this is a transient window.
     */
    if (CheckFlags(TRANSIENT) && qvMain) {
      List<Qvwm>::Iterator li(&qvMain->trList);
      Qvwm* qvWm = li.GetHead();

      while (qvWm) {
	if (qvWm == this) {
	  qvMain->trList.Remove(this);
	  break;
	}
	qvWm = li.GetNext();
      }
    }
    else {
      /*
       * When this is a main window which has some transient windows.
       */
      List<Qvwm>::Iterator li(&trList);
      Qvwm* qvTrans = li.GetHead();

      while (qvTrans) {
	qvTrans->qvMain = NULL;
	qvTrans = li.GetNext();
      }
    }

    if (CheckFlags(ONTOP))
      desktop.GetOnTopList().Remove(this);

    if (UsePager)
      delete mini;

    XDestroyWindow(display, frame);
  }
  else
    delete ctrlMenu;

  delete [] name;
  delete [] shortname;
}

void Qvwm::GetWindowAttrs(Bool usrPos)
{
  XWindowAttributes attr;

  XGetWindowAttributes(display, wOrig, &attr);

  /*
   * Preserve original border width and set border width to 0.
   */
  bwOrig = attr.border_width;
  XSetWindowBorderWidth(display, wOrig, 0);

#ifdef USE_XSMP
  /*
   * Restore properties and attributes from the previous session
   */
  GetSMClientId(); // initialize smClientId
  GetWMRole(); // initialize wmRole
  int ax, ay, aw, ah;
  if (session->SearchProto(smClientId, wmRole,
			   classHints.res_class,
			   classHints.res_name,
			   name, &ax, &ay, &aw, &ah)) {
    usrPos = True; // force place
    attr.x = ax;
    attr.y = ay;
    attr.width = aw;
    attr.height = ah;
  }
#endif // USE_XSMP

  Point pageoff;
  ExecPending* pending = ExecPending::LookInList(name, classHints);

  if (pending) {
    pageoff = pending->GetPageOff();
    ExecPending::m_pendingList.Remove(pending);
    delete pending;
  }
  else
    pageoff = paging->origin;

  /*
   * Calculate the window position.
   */
  if (usrPos || (flags & (TRANSIENT | FIXED_POS)) ||
      (hints.flags & (USPosition | PPosition))) {
    if (CheckFlags(GEOMETRY)) {
      if (geom->bitmask & WidthValue)
	attr.width = geom->rc.width;
      if (geom->bitmask & HeightValue)
	attr.height = geom->rc.height;
    }

    if (CheckFlags(STICKY))
      rcOrig = Rect(attr.x, attr.y, attr.width, attr.height);
    else
      rcOrig = Rect(attr.x + pageoff.x, attr.y + pageoff.y,
		    attr.width, attr.height);

    CalcWindowPos(rcOrig);
  }
  else {
    if (CheckFlags(GEOMETRY)) {
      if (geom->bitmask & WidthValue)
	attr.width = geom->rc.width;
      if (geom->bitmask & HeightValue)
	attr.height = geom->rc.height;

      if (geom->bitmask & XNegative)
	attr.x = DisplayWidth(display, screen) + geom->rc.x - attr.width;
      else if (geom->bitmask & XValue)
	attr.x = geom->rc.x;

      if (geom->bitmask & YNegative)
	attr.y = DisplayHeight(display, screen) + geom->rc.y - attr.height;
      else if (geom->bitmask & YValue)
	attr.y = geom->rc.y;

      if (CheckFlags(STICKY))
	rcOrig = Rect(attr.x, attr.y, attr.width, attr.height);
      else
	rcOrig = Rect(attr.x + pageoff.x, attr.y + pageoff.y,
		      attr.width, attr.height);
      
      CalcWindowPos(rcOrig);
    }
    else {
      int borderWidth, topBorder, titleHeight, titleEdge;

      GetBorderAndTitle(borderWidth, topBorder, titleHeight, titleEdge);

      rc.width = attr.width + borderWidth * 2;
      rc.height = attr.height + borderWidth + topBorder + titleHeight
	+ titleEdge;

      Point pt = desktop.GetNextPlace(Dim(rc.width, rc.height), pageoff);

      rc.x = pt.x;
      rc.y = pt.y;

      if (CheckFlags(STICKY))
	rcOrig = Rect(rc.x - rcScreen.x - pageoff.x,
		      rc.y - rcScreen.y - pageoff.y,
		      attr.width,
		      attr.height);
      else
	rcOrig = Rect(rc.x - rcScreen.x, rc.y - rcScreen.y,
		      attr.width, attr.height);
    }
  }
}

/*
 * ChangeOrigWindow --
 *   Change parent and attributes of original window.
 */
void Qvwm::ChangeOrigWindow()
{
  XSetWindowAttributes attributes;
  unsigned long valueMask;

  XReparentWindow(display, wOrig, parent, 0, 0);

  attributes.event_mask = StructureNotifyMask | PropertyChangeMask |
    EnterWindowMask | LeaveWindowMask | ColormapChangeMask;

#if defined(USE_SS) && !defined(USE_SSEXT)
  XWindowAttributes attr;

  XGetWindowAttributes(display, wOrig, &attr);
  attributes.event_mask |= attr.your_event_mask;
#endif

  attributes.do_not_propagate_mask = ButtonPressMask | ButtonReleaseMask;
  valueMask = CWEventMask | CWDontPropagate;

  XChangeWindowAttributes(display, wOrig, valueMask, &attributes);

  XSaveContext(display, wOrig, context, (caddr_t)this);
  XAddToSaveSet(display, wOrig);

#if defined(USE_SS) && !defined(USE_SSEXT)
  // check KeyPressMask of subwindows of wOrig again
  scrSaver->SelectEvents(wOrig);
#endif
}

/*
 * SetTransient
 *   Search pre-mapped transient windows for this window to handle
 *   a transient window properly even if the main window is mapped
 *   after the transient window is mapped first.
 */
void Qvwm::SetTransient()
{
  List<Qvwm>::Iterator i(&desktop.GetQvwmList());
  Window tWin;

  for (Qvwm* qvWm = i.GetHead(); qvWm; qvWm = i.GetNext()) {
    if (qvWm->CheckFlags(TRANSIENT) && qvWm->qvMain == NULL)
      if (XGetTransientForHint(display, qvWm->GetWin(), &tWin))
	if (tWin == wOrig) {
	  qvWm->qvMain = this;  // qvWm is a transient window of this
	  if (CheckFlags(ONTOP) && !qvWm->CheckFlags(ONTOP)) {
	    qvWm->SetFlags(ONTOP);
	    desktop.GetOnTopList().InsertTail(qvWm);
	    Gnome::ChangeLayer(qvWm, WIN_LAYER_ONTOP);
	  }
	  trList.InsertTail(qvWm);
	}
  }
}

/*
 * SendMessage --
 *   Send the messeage to this window.
 */
void Qvwm::SendMessage(Atom atom)
{
  XClientMessageEvent cev;

  cev.type = ClientMessage;
  cev.window = wOrig;
  cev.message_type = _XA_WM_PROTOCOLS;
  cev.format = 32;
  cev.data.l[0] = atom;
  cev.data.l[1] = CurrentTime;

#ifdef DEBUG
  if (atom == _XA_WM_DELETE_WINDOW)
    printf("Send delete msg to '%s'\n", name);
#endif

  XSendEvent(display, wOrig, False, 0L, (XEvent *)&cev);
}

/*
 * AdjustPage --
 *   Switch to the page where the window center point exists.
 */
void Qvwm::AdjustPage()
{
  if (this == rootQvwm)
    return;

  if (CheckFlags(STICKY))
    return;

  Rect rect = GetRect();
  Point ptCenter(rect.x+rect.width/2, rect.y+rect.height/2);
  Rect rcRoot = rootQvwm->GetRect();
  
  if (ptCenter.x >= 0)
    ptCenter.x =  ptCenter.x / rcRoot.width * rcRoot.width;
  else
    ptCenter.x = (ptCenter.x / rcRoot.width - 1) * rcRoot.width;

  if (ptCenter.y >= 0)
    ptCenter.y = ptCenter.y / rcRoot.height * rcRoot.height;
  else 
    ptCenter.y = (ptCenter.y / rcRoot.height - 1) * rcRoot.height;
  
  ASSERT(paging);

  paging->PagingProc(ptCenter, False);
}

/*
 * Exposure --
 *   Process exposure event.
 */
void Qvwm::Exposure(Window win)
{
  if (win == ctrl)
    DrawCtrlMenuMark();
  else if (win == title)
    DrawTitle(False);
  else
    DrawFrame(win);  // win may not be a part of frame windows
}

/*
 * Button1Press --
 *   Process the press of button1(mouse left button).
 */
void Qvwm::Button1Press(Window win, Time clickTime, const Point& ptRoot,
			unsigned int state)
{
  if (this != rootQvwm)
    toolTip->Disable();

  if (UseTaskbar && TaskbarAutoHide && !taskBar->IsHiding())
    taskBar->HideTaskbar();

  if (win == ctrl)
    CtrlButton1Press(clickTime, ptRoot, state);
  else {
    Menu::UnmapAllMenus();
    
    if (!CheckFocus())
      SetFocus();

    if (Icon::focusIcon != NULL)
      Icon::focusIcon->ResetFocus();

    if (ClickingRaises || win != wOrig)
      RaiseWindow(True);

    /*
     * Title bar.
     */
    if (win == title)
      TitleButton1Press(clickTime, ptRoot);

    SetStatus(PRESS_FRAME);
  }

  if (AutoRaise && focusQvwm == this && activeQvwm != this)
    RaiseWindow(True);
}

/*
 * Button1Release --
 *   Process the release of button1(mouse left button).
 */
void Qvwm::Button1Release()
{
  ResetStatus(PRESS_FRAME);

  /*
   * Unmap control menu if the menu is shown and menu button is pressed.
   */
  if (umReserved) {
    ASSERT(ctrlMenu);
    ctrlMenu->UnmapMenu();
    umReserved = False;
  }
}    

void Qvwm::Button2Press(Window win)
{
  if (this != rootQvwm)
    toolTip->Disable();

  if (win == side[0] || win == side[1] || win == side[2] || win == side[3] ||
      win == corner[0] || win == corner[1] || win == corner[2] ||
      win == corner[3])
    SetStatus(PRESS_FRAME);

  if (AutoRaise && focusQvwm == this && activeQvwm != this)
    RaiseWindow(True);
}

void Qvwm::Button2Release(Window win)
{
  ResetStatus(PRESS_FRAME);
  if (win != frame)
    LowerWindow();
}

void Qvwm::Button3Press()
{
  if (this != rootQvwm)
    toolTip->Disable();

  if (AutoRaise && focusQvwm == this && activeQvwm != this)
    RaiseWindow(True);
}

/*
 * Button3Release --
 *   Process the release of button3(mouse right button).
 */
void Qvwm::Button3Release(Window win, const Point& ptRoot)
{
  Point pt;
  int dir;

  if (this == rootQvwm) {
    Menu::UnmapAllMenus();
    
    if (!CheckFocus())
      SetFocus();
    
    ASSERT(ctrlMenu);
    pt = ctrlMenu->GetFixedMenuPos(ptRoot, dir);
    ctrlMenu->MapMenu(pt.x, pt.y, dir);
  }
  else {
    Menu::UnmapAllMenus();
    
    if (!CheckFocus())
      SetFocus();

    if (CheckMapped())
      RaiseWindow(True);
    else
      RestoreWindow();

    ASSERT(ctrlMenu);
    ctrlMenu->SetQvwm(this);
    pt = ctrlMenu->GetFixedMenuPos(ptRoot, dir);
    ctrlMenu->MapMenu(pt.x, pt.y, dir);
  }
}

/*
 * Button1Motion --
 *   Move or resize according to mouse movement.
 */
void Qvwm::Button1Motion(Window win, const Point& ptRoot)
{
  if (win == title && !CheckStatus(MAXIMIZE_WINDOW))
    MoveWindow(ptRoot, True);
  else if (win == side[0])
    ResizeWindow(P_LEFT, True);
  else if (win == side[1])
    ResizeWindow(P_RIGHT, True);
  else if (win == side[2])
    ResizeWindow(P_TOP, True);
  else if (win == side[3])
    ResizeWindow(P_BOTTOM, True);
  else if (win == corner[0])
    ResizeWindow(P_LEFT | P_TOP, True);
  else if (win == corner[1])
    ResizeWindow(P_RIGHT | P_TOP, True);
  else if (win == corner[2])
    ResizeWindow(P_LEFT | P_BOTTOM, True);
  else if (win == corner[3])
    ResizeWindow(P_RIGHT | P_BOTTOM, True);
}

void Qvwm::Button2Motion(Window win, const Point& ptRoot)
{
  if (!CheckStatus(MAXIMIZE_WINDOW))
    MoveWindow(ptRoot, True);
}

void Qvwm::Enter(Window win, unsigned int state)
{
  if (win == title) {
    if (strcmp(name, shortname) != 0)
      toolTip->SetTimer();
  }

  if ((state & NoFocusChangeMask) == NoFocusChangeMask)
    return;

  if (!ClickToFocus && !(NoDesktopFocus && this == rootQvwm) &&
      !CheckStatus(MINIMIZE_WINDOW) && !CheckFlags(FOCUS_ON_CLICK)) {
    if (!CheckFocus() && !Menu::CheckAnyMenusMapped()) {
      if (AutoRaise)
	RaiseWindow(False);
      SetFocus();
    }
  }
}

void Qvwm::Leave(Window win)
{
  if (win == title)
    toolTip->Disable();
}

void Qvwm::PointerMotion(Window win)
{
  if (win == title) {
    if (!toolTip->IsMapped())
      toolTip->ResetTimer();
  }
}

void Qvwm::Initialize()
{
  context = XUniqueContext();

  TITLE_HEIGHT = FrameTitleHeight;
  BORDER_WIDTH = FrameBorderWidth + 4;
  TOP_BORDER = FrameBorderWidth + 2;

  CreateTitlebarPixmap();
  CreateFramePixmap();
}

#ifdef USE_XSMP
/*
 * SavePropertiesToFile --
 *   Save the state of this window to the FILE for XSMP
 */
Bool Qvwm::SavePropertiesToFile(int idx, FILE* fp)
{
  if (qvMain) {
    // this window is a transient win.
    return True;
  }
  GetSMClientId();
  GetWMRole();
  // WM_CLASS  (if WM_WINDOW_ROLE is not specified)
  XClassHint wm_class;        // STRING
  if (wmRole == NULL) {
    if (XGetClassHint(display, wOrig, &wm_class) == 0)
      return False;
  }

  // Write out to the 'fp'
  if (smClientId != NULL)
    fprintf(fp, "%d,id=%s\n", idx, smClientId);       // SM_CLIENT_ID

  fprintf(fp, "%d,wOrig=%x\n", idx, (int)wOrig);
  fprintf(fp, "%d,frame=%x\n", idx, (int)frame);
  if (wmRole)
    fprintf(fp, "%d,role=%s\n", idx, wmRole); // WM_WINDOW_ROLE
  else {
    fprintf(fp, "%d,class=%s\n", idx, wm_class.res_class);// WM_CLASS
    fprintf(fp, "%d,rname=%s\n", idx, wm_class.res_name);
    fprintf(fp, "%d,name=%s\n", idx, name);   // WM_NAME
  }
  fprintf(fp, "%d,pos=%d,%d\n", idx, rcOrig.x, rcOrig.y);
  fprintf(fp, "%d,size=%d,%d\n", idx, rcOrig.width, rcOrig.height);

  if (wmRole == NULL) {
    XFree(wm_class.res_class);
    XFree(wm_class.res_name);
  }
  return True;
}
#endif // USE_XSMP
